### PR TITLE
[Feature] Legal hold controller implementation

### PR DESCRIPTION
**Pull request by johnxnguyen**
## What's new in this PR?

This PR fills in the `LegalHoldController` stubs to provide functionally to

- determine if legal hold is active for a particular user
- determine if legal hold is active in a particular conversation
- get a sequence  of all legal hold subjects in a particular conversation
#### APK
[Download build #3382](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3382/artifact/build/artifact/wire-dev-PR3270-3382.apk)
[Download build #3385](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3385/artifact/build/artifact/wire-dev-PR3270-3385.apk)
[Download build #3386](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3386/artifact/build/artifact/wire-dev-PR3270-3386.apk)
[Download build #3395](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3395/artifact/build/artifact/wire-dev-PR3270-3395.apk)
[Download build #3396](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3396/artifact/build/artifact/wire-dev-PR3270-3396.apk)
[Download build #3407](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3407/artifact/build/artifact/wire-dev-PR3270-3407.apk)
[Download build #3438](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3438/artifact/build/artifact/wire-dev-PR3270-3438.apk)
[Download build #3439](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3439/artifact/build/artifact/wire-dev-PR3270-3439.apk)

### CODE REVIEW

**Review by makingthematrix — CHANGES_REQUESTED** [review]
I'd propose that instead of injecting `Signal[*Storage]` here directly, and then working on those signals in a complicated way, you put that logic in `OtrClientsService`, `ConversationService`, and `UserService` respectively. There, you will be able to work on just `*Storage` instead of `Signal[*Storage]`, and here, in the controller, you will just access `Signal[*Service]` and `flatMap` it to a respective method.

In general, if a lot of work in a controller is done on signals of storages, it's a good hint that it should be moved to the service layer.

**Comment by gizemb** [comment]
> I'd propose that instead of injecting `Signal[*Storage]` here directly, and then working on those signals in a complicated way, you put that logic in `OtrClientsService`, `ConversationService`, and `UserService` respectively. There, you will be able to work on just `*Storage` instead of `Signal[*Storage]`, and here, in the controller, you will just access `Signal[*Service]` and `flatMap` it to a respective method.
> 
> In general, if a lot of work in a controller is done on signals of storages, it's a good hint that it should be moved to the service layer.

I agree that we should move all this logic to the Service layer. We can directly handle this in `LegalHoldService`, in order to keep all the LH logic in a single place.

**Comment by johnxnguyen** [comment]
@makingthematrix and @gizemb I moved the logic out of the controller as requested, and additionally added tests.

File: zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala [hunk 1]
```diff
@@ -28,16 +29,145 @@ class LegalHoldServiceSpec extends AndroidFreeSpec {
...
+      val actualResult = result(service.isLegalHoldActive(convId).future)
+
+      // Then
+      actualResult shouldBe false
+    }
+  }
+
+  feature("Legal hold users in a conversation") {
+
+    scenario("that contains legal hold subjects") {
```
Comment (gizemb) [inline zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala:146]: what about "that does not contain any legal hold subjects" case?

File: zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala [hunk 2]
```diff
@@ -28,16 +29,145 @@ class LegalHoldServiceSpec extends AndroidFreeSpec {
...
+      membersStorage,
+      cryptoSessionService
+    )
+
     userPrefs.setValue(UserPreferences.LegalHoldRequest, None)
   }
 
+  // Helpers
+
+  def expectation(userId: UserId, deviceTypes: Seq[OtrClientType]): Unit = {
```
[the remark this answers is not in the document]
Comment (johnxnguyen) [reply zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala]: Fixed in a3a8a5bf1331237019ac4a445bbde8cef748bea9

File: zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala [hunk 3]
```diff
@@ -28,16 +29,145 @@ class LegalHoldServiceSpec extends AndroidFreeSpec {
...
+      clientId -> Client(clientId, "", devType = deviceType)
+    }
+
+    (clientsStorage.optSignal _)
+      .expects(userId)
+      .once()
+      .returning(Signal.const(Some(UserClients(userId, clients.toMap))))
+  }
+
+  def expectation(convId: ConvId, legalHoldStatus: LegalHoldStatus): Unit =
```
[the remark this answers is not in the document]
Comment (johnxnguyen) [reply zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala]: Fixed in a3a8a5bf1331237019ac4a445bbde8cef748bea9

File: zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala (hunk 1 above)
Comment (johnxnguyen) [reply zmessaging/src/test/scala/com/waz/service/LegalHoldServiceSpec.scala:146]: Added in 671b99308ee6c8011aa20b146bc468725507f4d6

**Review by gizemb — APPROVED** [review]

File: zmessaging/src/main/scala/com/waz/service/LegalHoldService.scala [hunk 4]
```diff
@@ -29,23 +32,38 @@ class LegalHoldServiceImpl(selfUserId: UserId,
...
+  def isLegalHoldActive(userId: UserId): Signal[Boolean] =
+    clientsStorage.optSignal(userId).map(_.fold(false)(_.clients.values.exists(_.isLegalHoldDevice)))
+
+  def isLegalHoldActive(conversationId: ConvId): Signal[Boolean] =
+    convsStorage.optSignal(conversationId).map(_.fold(false)(_.isUnderLegalHold))
+
+  def legalHoldUsers(conversationId: ConvId): Signal[Seq[UserId]] = for {
+    users             <- membersStorage.activeMembers(conversationId)
+    usersAndStatus    <- Signal.sequence(users.map(userId => Signal.zip(Signal.const(userId), isLegalHoldActive(userId))).toSeq: _*)
+    legalHoldSubjects = usersAndStatus.filter(_._2).map(_._1)
```
[the remark this answers is not in the document]
Comment (makingthematrix) [reply zmessaging/src/main/scala/com/waz/service/LegalHoldService.scala:59]: Maybe something like this?
```
usersAndStatus    <- Signal.sequence(users.map(userId => isLegalHoldActive(userId).map(active => userId -> active)))
legalHoldSubjects =  usersAndStatus.collect { case (userId, true) => userId }
```

**Review by makingthematrix — APPROVED** [review]

**Pull request merged**